ui/tests/integration/toggle-test.js

```diff
@@ -0,0 +1,89 @@
+import { find, settled } from '@ember/test-helpers';
+import { module, test } from 'qunit';
+import { setupRenderingTest } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+import sinon from 'sinon';
+import { create } from 'ember-cli-page-object';
+import togglePageObject from 'nomad-ui/tests/pages/components/toggle';
+
+const Toggle = create(togglePageObject());
+
+module('Integration | Component | toggle', function(hooks) {
+  setupRenderingTest(hooks);
+
+  const commonProperties = () => ({
+    isActive: false,
+    isDisabled: false,
+    label: 'Label',
+    onToggle: sinon.spy(),
+  });
+
+  const commonTemplate = hbs`
+    {{#toggle
+      isActive=isActive
+      isDisabled=isDisabled
+      onToggle=onToggle}}
+      {{label}}
+    {{/toggle}}
+  `;
+
+  test('presents as a label with an inner checkbox and display span, and text', async function(assert) {
+    const props = commonProperties();
+    this.setProperties(props);
+    await this.render(commonTemplate);
+
+    assert.equal(Toggle.label, props.label, `Label should be ${props.label}`);
+    assert.ok(Toggle.isPresent);
+    assert.notOk(Toggle.isActive);
+    assert.ok(find('[data-test-toggler]'));
+    assert.equal(
+      find('[data-test-input]').tagName.toLowerCase(),
+      'input',
+      'The input is a real HTML input'
+    );
+    assert.equal(
+      find('[data-test-input]').getAttribute('type'),
+      'checkbox',
+      'The input type is checkbox'
+    );
+  });
+
+  test('the isActive property dictates the active state and class', async function(assert) {
+    const props = commonProperties();
+    this.setProperties(props);
+    await this.render(commonTemplate);
+
+    assert.notOk(Toggle.isActive);
+    assert.notOk(Toggle.hasActiveClass);
+
+    this.set('isActive', true);
+    await settled();
+
+    assert.ok(Toggle.isActive);
+    assert.ok(Toggle.hasActiveClass);
+  });
+
+  test('the isDisabled property dictates the disabled state and class', async function(assert) {
+    const props = commonProperties();
+    this.setProperties(props);
+    await this.render(commonTemplate);
+
+    assert.notOk(Toggle.isDisabled);
+    assert.notOk(Toggle.hasDisabledClass);
+
+    this.set('isDisabled', true);
+    await settled();
+
+    assert.ok(Toggle.isDisabled);
+    assert.ok(Toggle.hasDisabledClass);
+  });
+
+  test('toggling the input calls the onToggle action', async function(assert) {
+    const props = commonProperties();
+    this.setProperties(props);
+    await this.render(commonTemplate);
+
+    await Toggle.toggle();
+    assert.equal(props.onToggle.callCount, 1);
+  });
```

Comment:
In the vein of the above comment on keyboard access, I tried it out and it all worked as expected (apart from a moment of 🤔 where triggering “Drain” with the keyboard happened too quickly for the progress indicator to even show that I’d actually done it successfully, but I think that’s something that won’t happen in a real scenario vs local testing), but what do you think of having an assertion for that at this level too? I find it somewhat unfortunate to having a multiplication of tests for different triggers but I also like to have some assertions that catch possible regressions… so maybe this could be `Toggle.click` and then `Toggle.spaceBar` in another test, or the like?

Reply:
I had that at one point, but then I couldn't get it to work. Turns out the spacebar action on inputs actually invoke the `click` handler. 

So I agree that it would be good to capture, since it's not hard to imagine the checkbox being replaced with a `span` or something at some point and we accidentally lose the keyboard support, but I'm not sure how to mock it since it isn't actually a different event.

Reply:
I’ve tested keyboard things with page objects before; I haven’t tried it in this situation but maybe this would work:

```
spaceBar: triggerable('keypress', '[data-test-input]', { eventProperties: { keyCode: 32 } })
```

Reply:
The issue isn't emulating the key event. It's that emulating the key event doesn't toggle the checkbox. 

I don't know what the browser is doing, but when you press space when the checkbox is active, it sets `checked` and fires a `click` event.

When you emulate the space press, nothing happens. It's as if key controls for form inputs is a browser detail that gets translated into different DOM events before making it into javascript land.

I pushed up a commit to show you what I tried: https://github.com/hashicorp/nomad/commit/050f60a1baf7b8954325ba18500eebb599147d1b

Reply:
oh wow, well that’s disappointing, sorry for misunderstanding 😞